### PR TITLE
Add virustotal check for casks

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Checks virustotal for positive results matching the download URL.

Fixes https://github.com/Homebrew/homebrew-cask/issues/56223

### CODE REVIEW

**Comment by vitorgalvao** [comment]
So this works only if a user has an API key and sets it in `HOMEBREW_VIRUSTOTAL_API_TOKEN`, correct?

Does this API call cause VirusTotal to download the app on their end? I ask because on their website, giving a URL is not the same as giving a file upload. A URL is detected as malware by fewer engines than uploading the download of that URL.

Pinging @core-code and @randall.

**Comment by SMillerDev** [comment]
Yes, you will need to set an api key.
It also only checks a URL. It's not a perfect solution but it'll already catch more than without the audit. 

I also still want to make virustotal scan the URL of it can't find results but I only thought of that once I read your comments just now.

File: Library/Homebrew/cask/audit.rb [hunk 1]
```diff
@@ -6,6 +6,7 @@
 require "digest"
 require "utils/curl"
 require "utils/git"
+require "extend/ENV"
```
Comment (reitermarkus) [inline Library/Homebrew/cask/audit.rb:9]: Why is this needed now?

Comment (SMillerDev) [reply Library/Homebrew/cask/audit.rb:9]: Isn't that what allows me to read ENV?

Comment (MikeMcQuaid) [reply Library/Homebrew/cask/audit.rb:9]: Nah, can do so unconditionally.

File: Library/Homebrew/cask/audit.rb [hunk 2]
```diff
@@ -322,6 +324,28 @@ def check_appcast_contains_version
       add_error "appcast at URL '#{appcast_stanza}' offline or looping"
     end
 
+    def check_virustotal
+      return unless ENV["HOMEBREW_VIRUSTOTAL_API_TOKEN"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:328]:
```suggestion
      return if ENV["HOMEBREW_VIRUSTOTAL_API_TOKEN"].blank?
```

File: Library/Homebrew/cask/audit.rb [hunk 3]
```diff
@@ -322,6 +324,28 @@ def check_appcast_contains_version
...
+      api_url = "https://www.virustotal.com/vtapi/v2/url/report?apikey=#{token}&scan=1&resource=#{cask.url}"
+      out, _, status= curl_output("--request", "GET", api_url)
+      return unless status.success?
+
+      metadata = JSON.parse(out)
+      return if metadata.nil?
+
+      ohai "Scan in progress, result at #{metadata["permalink"]}" if metadata["response_code"] == -2
+
+      return unless metadata["response_code"] == 1
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:342]:
```suggestion
      return if metadata["response_code"] != 1
```

**Comment by core-code** [comment]
@SMillerDev thanks for working on improvements here, happy to see that!

i've been running regular VirusTotal checks over all Casks for well over a year and i think i can confirm what @vitorgalvao has alluded to above: checking the URL itself is not really useful. i am not even sure it qualifies as 'better than nothing'. what you need to do is an actual file check. there are two ways to do it:

• check the SHA256 given in the CASK or

• do an URL check first and then look at the 'filescan_id' and then do another 'file' lookup with this ID

i've been doing both, since you never know .... ;-) 

additionally you often get 'old' results from VirusTotal so i'm also using their 'rescan' API to initiate a re-check for every result older than 8 weeks.


i'd be happy to share our code for this if this would help. its around 250 lines of ObjC

**Comment by MikeMcQuaid** [comment]
> i'd be happy to share our code for this if this would help. its around 250 lines of ObjC

It would be useful to have this in Homebrew/brew. For that, however, it'll need to be ported to Ruby.

**Comment by core-code** [comment]
i can provide the code, and could even work on enabling it to run in a more stand-alone fashion if that is required, but sorry i can't port it to Ruby. 

but the concepts should be easy to see from it so that may help with getting the file-checks running. 

i've attached the relevant excerpt for reference.


[vt.m.txt](https://github.com/Homebrew/brew/files/4529038/vt.m.txt)

**Comment by SMillerDev** [comment]
I've updated the PR. I think anything that can potentially catch something is worth merging though. It might not catch as much as a full file upload but it's a quick check that has the possibility to catch issues.

**Comment by core-code** [comment]
> full file upload 

as mentioned there is no need upload anything. just check the SHA or the 'filescan_id' attached to the URL check response.

**Comment by ran-dall** [comment]
@vitorgalvao Thanks for the ping! I've had a rough couple of days fixing things for folk.

So yes, the ML already runs intended updates through [Malice](https://github.com/maliceio/malice) before running `cask-repair`; which was a recommendation by another HBC user (I forget who it was exactly). I thought Malice was/is a better fit for the ML as there is some wording in the VirusTotal agreement that caused me some concern. Now, Malice has some real problems, but I believe it does parse VirusTotal (and a handful or two of all the other APIs) successfully still. I meant to update Malice to fix some of the problems, but I haven't had the time to in the last few months. 

That said, per the estimations and analytics of the ML, the ML covers about 72% of the all the casks offered on HBC (with an error / fail rate of just under 8% [credit: @suschizu and @core-code]). So I would say it's probably not super necessary, as it's only going to get better on the ML's end as we continue. 

However, from an end-user perspective, I do agree with what @SMillerDev Is doing. Furthermore, I believe this would fit the VirusTotal user agreement because it's on the end-users' end.

**Comment by SMillerDev** [comment]
It could also be run by CI. Either way I believe that maintainers should not run anything on their machines to maintain parts of homebrew. So until the more advanced checks are submitted as audits, I think this will adds value.

**Comment by ran-dall** [comment]
> Either way, I believe that maintainers should not run anything on their machines to maintain parts of homebrew.

@SMillerDev To clarify, the ML is its own free-standing entity and the idea of its telemetry processes is to prevent bad updates from happening. Currently, if the telemetry raises a concern about an update; intervention is required. That said, the ML could probably run on its own without any human intervention; but most of those features are off and not really developed past a basic prototyping point.

**Comment by MikeMcQuaid** [comment]
> i can provide the code, and could even work on enabling it to run in a more stand-alone fashion if that is required, but sorry i can't port it to Ruby.
> 
> but the concepts should be easy to see from it so that may help with getting the file-checks running.

@core-code Yes, that would be good to have in a public repo somewhere, thanks.

**Comment by core-code** [comment]
@MikeMcQuaid here you are: https://github.com/core-code/VirusCheckHBC
